### PR TITLE
Overhaul doc_links testing infra

### LINKED ISSUE #9617: Goto definition on rustdoc [`T::links`] targets T rather than the target item
If you goto definition a link that refers to an item on a trait (maybe also structs/impls), it will go to the trait rather than the item. It would be more useful if it went to the specific item, especially on long definitions.

Another bug I found while writing this one up: If you replace `T` with `Self` in this link, goto definition won't work, but rustdoc will generate the links fine.

```rust
trait T {
   // ^ actual target
    /// Does the thing using [`do_part_of_the_thing`](T::do_part_of_the_thing).
                                           // ^- goto definition here
    /// This one also does the same thing [`T::do_part_of_the_thing`]
    fn do_thing();

    fn do_part_of_the_thing();
    // ^^^^^^^^^^^^^^^^^^^^ expected target
}

fn main() {}
```

Version: a2f83c9

### PR DESCRIPTION
**Pull request by Veykril**
and fix several issues with current implementation.

Fixes #9617

### CODE REVIEW

**Comment by Veykril** [comment]
Looks like this is gonna become a bigger PR, I just realized how many holes our docs related stuff has 😅

**Comment by matklad** [comment]
Yeah, I am not to happy about doc infra, it can use some spring cleaning!

File: crates/ide/src/doc_links.rs [hunk 1]
```diff
@@ -669,7 +635,185 @@ fn foo() {
...
+
+    fn def_under_cursor(
+        sema: &Semantics<RootDatabase>,
+        position: &FilePosition,
+    ) -> (Definition, hir::Documentation) {
+        let (docs, def) = sema
+            .parse(position.file_id)
+            .syntax()
+            .token_at_offset(position.offset)
+            .next()
```
Comment (matklad) [inline crates/ide/src/doc_links.rs:783]: Better to use `.left_biased()` here.

File: crates/ide/src/doc_links.rs [hunk 2]
```diff
@@ -669,7 +635,185 @@ fn foo() {
...
+                ast::Variant(it)     => sema.to_def(&it).map(|def| (def.docs(sema.db), Definition::ModuleDef(hir::ModuleDef::Variant(def)))),
+                ast::Trait(it)       => sema.to_def(&it).map(|def| (def.docs(sema.db), Definition::ModuleDef(hir::ModuleDef::Trait(def)))),
+                ast::Static(it)      => sema.to_def(&it).map(|def| (def.docs(sema.db), Definition::ModuleDef(hir::ModuleDef::Static(def)))),
+                ast::Const(it)       => sema.to_def(&it).map(|def| (def.docs(sema.db), Definition::ModuleDef(hir::ModuleDef::Const(def)))),
+                ast::TypeAlias(it)   => sema.to_def(&it).map(|def| (def.docs(sema.db), Definition::ModuleDef(hir::ModuleDef::TypeAlias(def)))),
+                ast::Impl(it)        => sema.to_def(&it).map(|def| (def.docs(sema.db), Definition::SelfType(def))),
+                ast::RecordField(it) => sema.to_def(&it).map(|def| (def.docs(sema.db), Definition::Field(def))),
+                ast::TupleField(it)  => sema.to_def(&it).map(|def| (def.docs(sema.db), Definition::Field(def))),
+                ast::Macro(it)       => sema.to_def(&it).map(|def| (def.docs(sema.db), Definition::Macro(def))),
```
Comment (matklad) [inline crates/ide/src/doc_links.rs:813]: This is so much rust-analyzer :-)

Comment (matklad) [reply crates/ide/src/doc_links.rs:813]: At some point, it probably makes sense to collect all `match_ast` `match def` usages, and see if a better API is possible.

**Pull request merged**